File: 05-binary_search/00704-Binary_Search/00704-binary_search.cpp
// 704: Binary Search
// https://leetcode.com/problems/binary-search/


#include <iostream>
#include <vector>

using namespace std;


class Solution {
public:
    // SOLUTION
    int search(vector<int>& nums, int target) {
        int l = 0;
        int h = nums.size() - 1;

        while (l <= h) {
            int m = l + (h-l)/2;
            
            if (nums[m] < target)
                l = m+1;
            else if (nums[m] > target)
                h = m-1;
            else
                return m;
        }

        return -1;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> nums = {-1,0,3,5,9,12};
    int target = 9;

    // OUTPUT
    auto result = o.search(nums, target);
    cout<<result<<endl;

    return 0;
}

File: 05-binary_search/00704-Binary_Search/00704-binary_search.py
# 704: Binary Search
# https://leetcode.com/problems/binary-search/


class Solution:
    # SOLUTION
    def search(self, nums: list[int], target: int) -> int:
        l, h = 0, len(nums) - 1

        while l <= h:
            m = l + (h-l)//2
            
            if nums[m] < target:
                l = m+1
            elif nums[m] > target:
                h = m-1
            else:
                return m

        return -1



if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums: list[int] = [-1,0,3,5,9,12]
    target: int = 9

    # OUTPUT
    result = o.search(nums, target)
    print(result)